sql/src/main/java/io/crate/operation/collect/collectors/ShardStateAwareRemoteCollector.java

```diff
@@ -0,0 +1,218 @@
+ * a relocation.
+ * In case the cluster state does not yet reflect the fact that the target shard was relocated (ie. it is closed locally
+ * but it's routing status is still RELOCATING) it will wait for the shard to get started and then start collecting
+ * (remote collecting in case the shard relocated to a different node or a local collection if the shard ended up back
+ * on the local node)
+ */
+public class ShardStateAwareRemoteCollector implements CrateCollector {
+
+    private static final Logger LOGGER = Loggers.getLogger(ShardStateAwareRemoteCollector.class);
+
+    private final String localNode;
+    private final RowConsumer consumer;
+    private final Object killLock = new Object();
+
+    private final ClusterService clusterService;
+    private final int shardId;
+    private final String indexName;
+    private final IndicesService indicesService;
+    private final Function<IndexShard, CrateCollector> localCollectorProvider;
+    private final Function<String, RemoteCollector> remoteCollectorProvider;
+    private final ExecutorService executorService;
+    private final ThreadContext threadContext;
+    private boolean collectorKilled = false;
+    private CrateCollector collector;
+    private TimeValue stateObserverWaitForChangeTimeout = new TimeValue(60000);
+
+    public ShardStateAwareRemoteCollector(String indexName,
+                                          int shardId,
+                                          RowConsumer consumer,
+                                          ClusterService clusterService,
+                                          IndicesService indicesService,
+                                          Function<IndexShard, CrateCollector> localCollectorProvider,
+                                          Function<String, RemoteCollector> remoteCollectorProvider,
+                                          ExecutorService executorService,
+                                          ThreadContext threadContext) {
+        this.indexName = indexName;
+        this.shardId = shardId;
+        this.consumer = consumer;
+        this.clusterService = clusterService;
+        this.localNode = clusterService.localNode().getId();
+        this.indicesService = indicesService;
+        this.localCollectorProvider = localCollectorProvider;
+        this.remoteCollectorProvider = remoteCollectorProvider;
+        this.executorService = executorService;
+        this.threadContext = threadContext;
+    }
+
+    @Override
+    public void doCollect() {
+        ClusterState clusterState = clusterService.state();
+        checkStateAndCollect(clusterState);
+    }
+
+    private void checkStateAndCollect(ClusterState clusterState) {
+        IndexShardRoutingTable shardRoutings;
+        try {
+            shardRoutings = clusterState.routingTable().shardRoutingTable(indexName, shardId);
+        } catch (ShardNotFoundException | IndexNotFoundException e) {
+            consumer.accept(null, e);
```

Comment:
Shouldn't we handle at least the `ShardNotFoundException` gracefully - in that we use the observer as well and wait for it to become available?